Fix workflow run details error

KaemonIsland opened the pull request:
## Hey, I just made a Pull Request!

The `WorkflowRunDetails` page was showing a `Failed to Build,` error for a split second when loading a workflow.

The issue is that the repo/owner variables are undefined during the first render of the component. This is causing the details variable to return an error stating that there is no repo/owner.

Once the correct repo/owner is set (Obtained from an async function) the details variable gets updated and returns the correct value.

I added a new check to the `loading` if statement that checks for the presence of `details.value` in addition to checking for an `error.message` considering `error.message` is used within the error message. (WOW, didn't mean to say error message 3(4) times)!

This screenshot displays the debugger image of the error message, and the current values of the WorkflowRunDetails variables.


![Screenshot 2022-02-04 133134](https://user-images.githubusercontent.com/37159349/152599059-fdb7da9f-ffb9-4d74-b326-5a048d44f994.png)

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

benjdlambert commented on plugins/github-actions/src/components/WorkflowRunDetails/WorkflowRunDetails.tsx:
```diff
@@ -175,14 +175,17 @@ export const WorkflowRunDetails = ({ entity }: { entity: Entity }) => {
   const jobs = useWorkflowRunJobs({ hostname, owner, repo });
 
   const error = projectName.error || (projectName.value && details.error);
```
I'm a little confused by the existing code, shouldn't it be something like:

```ts
const error = projectName.error ?? details.error;
const loading = projectName.loading ?? details.loading;
```

KaemonIsland replied:
The error and loading checks could stay the same as they are. The issue is due to asynchronous code that has `projectName.loading` and `details.loading` set to false, while the `details.error` was set because it was passed undefined values when `projectName` was still loading.

KaemonIsland replied:
I've explained the bug in more detail below.

Rugvip replied:
I feel the most straight-forward fix for this issue is to make the `useProjectName` synchronous. I don't see any reason for being async as is. This does also seem to be the case for at least a couple of the other CI/CD plugins btw. If the fix is simple enough, is it something you could apply to the other plugins as well where needed?

benjdlambert commented on plugins/github-actions/src/components/WorkflowRunDetails/WorkflowRunDetails.tsx:
```diff
@@ -175,14 +175,17 @@ export const WorkflowRunDetails = ({ entity }: { entity: Entity }) => {
   const classes = useStyles();
-  if (error) {
+
+  if (error && error.message) {
```
Is this the real issue here, that error is set to something, but it's false positive? It's something that isn't a real error?

Rugvip replied:
Tbh we should fix the `Promise.reject('...')` of strings being thrown around. Just like `throw`, promise rejections should always come in the form on an `Error`.

benjdlambert reviewed (commented):
Thanks for this - just got some questions about what is actually happening in the hooks and that's causing this bug to happen in the first place. Feels like there's some pretty strange logic in the first place, so just wanna clarify that first and work out what's happening.

KaemonIsland commented:
Sorry, I should have been more descriptive on the actual problem. The problem is because of asynchronous code. Although some objects are still loading, they are being passed to other hooks to get the rest of the necessary values, even though they are undefined.

## The Problem

So the cause of the bug is from line 167, `const projectName = useProjectName(entity);`

During the first render, `projectName` returns something along the lines of `{value: undefined, loading: true, error: undefined}`

The `projectName.value` get's split into owner/repo on line 173, `const [owner, repo] = projectName.value ? projectName.value.split('/') : [];`, which repo/owner are both undefined.

The undefined owner/repo get sent to useWorkflowRunsDetails on 157, `const details = useWorkflowRunsDetails({ hostname, owner, repo });`

Because owner/repo are undefined, useWorkflowRunsDetails returns `Promise.reject('No repo/owner provided');`

This is returned as `{error: 'No repo/owner provided'}`

At this point, `projectName` would be updated to its correct values containing the project information.

Once we reached the if statement we'd have these values:

1. Both `details.loading` and `projectName.loading` would be false.
2. `details` would have an error `Error: "No repo/owner provided`

The if statement sees that an error is present, (although it's not formatted as error.message), both loading checks are false, so we see "Failed to load build, " on the frontend.

## My Solution

The solution would be to wait until `projectName` was finished loading before we try to obtain the rest of the details. However, because `useWorkflowRunsDetails` is a hook, I can't call it from a `useEffect` as it'd break the rules of hooks.

I found that setting `loading` to `const loading = projectName.loading || details.loading || !details.value;` and the error check to `error && error.message` resolved this issue.

The second render resolves all issues as the projectName on 167 returns the project name.

Rugvip reviewed (commented):
Thank you! Sorry for letting this one sit for a bit.

KaemonIsland commented:
Okay, I've changed the `useProjectName` hook to be synchronous, which fixes the issue I was experiencing. I also updated all instances where it was used within the plugin and tested locally, no errors occurred.

I also updated the `Promise.reject('error string')` to return an `Error` object.

Rugvip commented on plugins/github-actions/src/components/WorkflowRunDetails/useWorkflowRunJobs.ts:
```diff
@@ -37,7 +37,7 @@ export const useWorkflowRunJobs = ({
           id: parseInt(id, 10),
         })
-      : Promise.reject('No repo/owner provided');
+      : Promise.reject(Error('No repo/owner provided'));
```
Preferred pattern is to use `new Error`, although I think the linter will complain about that too x)

KaemonIsland replied:
Done!

Rugvip commented on plugins/github-actions/src/components/useProjectName.ts:
```diff
@@ -13,15 +13,9 @@
-  });
-  return { value, loading, error };
-};
+export const useProjectName = (entity: Entity) =>
```
Tbh this is no longer a hook, so we could just call it `getProjectNameFromEntity` or something like that, to avoid confusion

KaemonIsland replied:
Good point! Done!

Rugvip reviewed (commented):
Nice, thank you! 👍 

Just some style nits left 😁

Rugvip reviewed (approved):
:shipit:
